Added condor_submit Quick Reference Guide (#905)
CHTC/chtc-website-source: _uw-research-computing/condor-submit-quick-ref.md

---
layout: guide
title: "Quick reference: HTCondor Submission commands"
alt_title: "Quick reference: HTCondor Submission commands"
guide:
    category: Get started
    tag: htc
---

## Introduction

This page lists common HTCondor commands and options for jobs. Users familiar with HTCondor and job submission on CHTC's High Throughput Computing (HTC) system can use this page as a quick reference. For users who are just starting out, we suggest reading the full guides (linked) to understand the full context of each command or option.

{% capture content %}
- [Introduction](#introduction)
- [Submit jobs](#submit-jobs)
- [Basic Submit File Options](#basic-submit-file-options)
- [Managing File Transfers in HTCondor](#managing-file-transfers-in-htcondor)
- [Controlling Where Your Job Runs](#controlling-where-your-job-runs)
- [Controlling How Your Job Runs](#controlling-how-your-job-runs)
- [Queue Statement Options](#queue-statement-options)
- [Glossary](#glossary)
- [Related Pages](#related-pages)
{% endcapture %}

> {% include /components/directory.html title="Table of Contents" %}

Reviewer: Update this after making changes.

> **Please note the following!** 
> 
> * Bracketed items (`<>`) denote where to place your input. 
>   Do not include the brackets in your command.
> * All commands should be entered on a single line.

> ## Submit jobs

Reviewer: ```suggestion
## Commands to submit jobs
```

| `condor_submit` | submits job(s) as specified by `submit_file` | See [monitor your jobs](condor_q) <br><br> **Example:** <br>`condor_submit helloWorld.sub`                                                                                                                                                |
| `condor_submit -i <submit_file>` | submits an interactive job as specified by `submit_file` | **Example:** <br>`condor_submit -i helloWorld.sub`                                                                                                                                                                                        |
| `condor_submit <submit_file> -append <attribute>` | appends Condor commands to `submit_file` during submission | Commands are added inplace prior to the `queue` command while leaving the `submit_file` unedited. Can pass `queue` commands as appended, if needed. <br><br> **Example:** <br>`condor_submit helloWorld.sub -append "request_disk = 1GB"` |
{:.command-table}

## Basic Submit File Options

> | Command | Use | Notes and Examples                                                                                                                                 |

Reviewer: ```suggestion
| Option | Use | Notes and Examples                                                                                                                                 |
```
Though they're called "commands" in the HTCondor manual, I would call these "options".

| --- | --- |----------------------------------------------------------------------------------------------------------------------------------------------------|

> | `executable = <executable.sh>` | path to the executable script | The executable script is automatically transferred to the Execution Point (EP) by HTCondor. <br><br> **Example:** <br>`executable = helloWorld.py` |

Reviewer: ```suggestion
| `executable = <executable.sh>` | specifies the path to the executable script | The executable script is automatically transferred to the Execution Point (EP) by HTCondor. <br><br> **Example:** <br>`executable = helloWorld.py` |
```

> | `arguments = "<args>"` | lists arguments to be passed to the executable as part of the command line | Arguments are wrapped by quotes(") and space separated. <br><br> **Example:** <br>`arguments = "hello world"`                                      |

Reviewer: Arguments do not have to be in quotes.

> | `environment = <parameter_list>` | lists environmental variables | Wrapped by quotes(") and space separated. <br><br> **Example:** <br>`environment = "VARIABLE1=Value1 VAR2='hello world'"`                          |

Reviewer: ```suggestion
| `environment = <parameter_list>` | sets additional environmental variables in the job environment | Wrapped by quotes(") and space separated. <br><br> **Example:** <br>`environment = "VARIABLE1=Value1 VAR2='hello world'"`                          |
```

I think this should be in a section called "Additional Options"

Author: a single line table _feels_ awkward. Should I move it anyways or drop the row entirely?

> | `log = <job.log>` | denotes the path to the log file | If not provided, no log file is recorded. <br><br> **Example:** <br>`log = ./log_files/job1.log`                                                   |

Reviewer: ```suggestion
| `log = <job.log>` | writes the HTCondor log file to this file | We always recommend specifying `log`. If `log` is not provided, no log file is written. <br><br> **Example:** <br>`log = ./log_files/job1.log`                                                   |
```

> | `output = <job.out>` | path to file capturing `stdout` screen output | Can be merged with `stderr` by denoting the same path in `error = <path>`. <br><br> **Example:** <br>`output = ./log_files/job1.out`               |

Reviewer: ```suggestion
| `output = <job.out>` | captures and writes `stdout` screen output to this file  | Can be merged with `stderr` by denoting the same path in `error = <path>`. <br><br> **Example:** <br>`output = ./log_files/job1.out`               |
```

> | `error = <job.err>` | path to file capturing `stderr` screen output | Can be merged with `stdout` by denoting the same path in `output = <path>`. <br><br> **Example:** <br>`error = ./log_files/job1.err`               |

Reviewer: ```suggestion
| `error = <job.err>` | captures and writes `stderr` screen output to this file | Can be merged with `stdout` by denoting the same path in `output = <path>`. <br><br> **Example:** <br>`error = ./log_files/job1.err`               |
```

{:.command-table}

> ## Managing File Transfers in HTCondor

Reviewer: ```suggestion
## File Transfer Options

Visit our [file transfer guide](htc-job-file-transfer) for more detail.
```

| Command | Use | Notes and Examples                                                                                                                                                                                                                                                             |
| --- | --- |--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `transfer_input_files = ` | lists all the input files to be transferred to the Execute Point (EP) | Comma-separated list. Various file transfer protocols can be used including `file:///`, `osdf:///`, `pelican:///`, and `s3:///`. <br><br> **Examples:** <br> `transfer_input_files = osdf:///chtc/staging/...`                                                                 |

> | `transfer_output_files = ` | explicitly lists the path to files to be returned to the Access Point (AP) | **Example:** <br><br>`transfer_output_files = ./results.txt`                                                                                                                                                                                                                   |

Reviewer: ```suggestion
| `transfer_output_files = ` | explicitly lists the path to files to be returned to the Access Point (AP) | If this is not specified, HTCondor will only transfer new and changed **files** in the top-level directory of the Execution Point. <br><br> **Example:** <br><br>`transfer_output_files = results.txt`                                                                                                                                                                                                                   |
```

> | `transfer_output_remaps = "<path_on_EP>=<new_path_on_AP>"` | remaps output files to a new location on the AP upon job completion | File paths on the left side (e.g., ./results.txt) refer to locations on the Execution Point (EP), while remapped paths on the right side refer to the Access Point (AP). <br><br> **Example:** <br>`transfer_output_remaps = "./results.txt=/staging/<user>/job1_results.txt"` |

Reviewer: ```suggestion
| `transfer_output_remaps = "<path_on_EP>=<new_path_on_AP>"` | remaps output files to a new location on the AP upon job completion | File paths on the left side (e.g., ./results.txt) refer to locations on the Execution Point (EP), while remapped paths on the right side refer to the Access Point (AP). <br><br> **Example:** <br>`transfer_output_remaps = "results.txt=/staging/<user>/job1_results.txt"` |
```

{:.command-table}

> ## Controlling Where Your Job Runs

Reviewer: ```suggestion
## Request Resources
```

| Command                               | Use                                                 | Notes and Examples                                                                                                                                                                       |
|---------------------------------------|-----------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `request_cpus = <int>`                | requests number of CPUs (cores)                     | **Example:** <br>`request_cpus = 4`                                                                                                                                                      |

> | `request_disk = <quantity>`           | requests disk space (Default in KB)                 | Can use units like `K`, `M`, `G`, or `T`. <br><br> **Example:** <br>`request_disk = 40GB`                                                                                                |

Reviewer: HTCondor manual says default is in KiB. https://htcondor.readthedocs.io/en/latest/man-pages/condor_submit.html#request_disk

> | `request_memory = <quantity>`         | requests memory for job (Default in KB)             | **Example:** <br>`request_memory = 250GB`                                                                                                                                                |

Reviewer: HTCondor manual says default is in MB. https://htcondor.readthedocs.io/en/latest/man-pages/condor_submit.html#request_memory

| `gpus_minimum_capability = <version>` | sets minimum GPU capability to specify the GPU type | **Example:** <br>`gpus_minimum_capability = 8.5`                                                                                                                                         |
| `gpus_maximum_capability = <version>` | sets maximum GPU capability to specify the GPU type | **Example:** <br>`gpus_maximum_capability = 9.0`                                                                                                                                         |
| `gpus_minimum_memory = <quantity in MB>` | specifies the minimum GPU VRAM memory to request | **Example:** <br>`gpus_minimum_memory = 3200`                                                                                                                                            |
| `want_campus_pools = True`            | Allows jobs to run on other HTCondor pools on campus (e.g., UW Grid) | Best for short (<8hr) or checkpointable jobs. Expands matching opportunities across campus. <br><br> **May not include `HasCHTCStaging` in the requirements**                            |

> | `want_ospool = True`                  | Allows jobs to match to the national Open Science Pool (OSPool) | Best for short (<8hr) or checkpointable jobs. Significantly increases available compute slots. <br><br> **May not include `HasCHTCStaging` in the requirements**                         |

Reviewer: I would put these two in a separate section, titled "Scale Beyond Local Capacity"

{:.command-table}

_Note: For more information on submitting GPU jobs, please visit our [Using GPUs](gpu-jobs.md) guide._

> ## Controlling How Your Job Runs

Reviewer: ```suggestion
## Specify the Container Software Environment
```

| Command | Use | Notes and Examples                                                                                                                |
| --- | --- |-----------------------------------------------------------------------------------------------------------------------------------|
| `container_image = <image_file>` | defines container image path | Can pull from DockerHub or use a local `.sif` file. <br><br> **Example:** <br>`container_image = docker://pytorch/pytorch:latest` |
{:.command-table}

> _Note: For more information on using containers in your jobs, please visit our [Using Apptainer Containers](apptainer-htc.md) guide._

Reviewer: ```suggestion
_Note: For more information on using containers in your jobs, please visit our [Using Apptainer Containers](apptainer-htc) or [Running HTC Jobs Using Docker Containers](docker-jobs) guide._
```

| `request_gpus = <integer>`                   | requests number of GPUs                                              | See our [GPU jobs guide](gpu-jobs). |
| `gpus_minimum_capability = <version>`    | sets minimum GPU capability               | **Example:** <br>`gpus_minimum_capability = 8.5`                                                                                                                                         |
| `gpus_maximum_capability = <version>`    | sets maximum GPU capability                | **Example:** <br>`gpus_maximum_capability = 9.0`                                                                                                                                         |
| `gpus_minimum_memory = <quantity>` | requests minimum GPU VRAM memory (Default in MB)                 | **Example:** <br>`gpus_minimum_memory = 3200`                                                                                                                                            |

> | `requirements = <ClassAd Boolean>`       | sets job execution constraints                                       | See [ClassAd reference](https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html). <br><br> **Example:** <br>`requirements = (OpSysAndVer == "RedHat9")` |

Author: ```suggestion
| `requirements = <ClassAd Boolean>`       | Specify job requirements to restrict jobs to certain execution points  | See [ClassAd reference](https://htcondor.readthedocs.io/en/lts/classad-attributes/machine-classad-attributes.html). <br><br> **Example:** <br>`requirements = (TARGET. HasCHTCStaging == true)` |
```
